### PR TITLE
Fix windowFunnel counting same event several times

### LINKED ISSUE #21835: windowFunnel: same event gets counted twice
(you don't have to strictly follow this form)

**Describe the unexpected behaviour**
Test 1: running `windowFunnel` on different events without duplicate timestamp, it works well:
```sql
SELECT
    uid,
    windowFunnel(50)(toUInt32(event_ms), event_id = 1, event_id = 2) AS level
FROM
    (
    /* emulate test dataset */
    SELECT data.1 event_id, data.2  event_ms, data.3 uid
    FROM (
        SELECT arrayJoin(
        [(1, 100, 123),
        (1, 200, 345),
        (1, 210, 345),
        (2, 99, 123),
        (2, 250, 234),
        (2, 230, 345),
        (2, 240, 345)]) data)
    )
WHERE
    event_ms >= 0
    AND event_ms <= 300
GROUP BY
    uid;

Results:
┌─uid─┬─level─┐
│ 123 │     1 │
│ 234 │     0 │
│ 345 │     2 │
└─────┴───────┘
```

Test 2: running `windowFunnel` on different events with duplicate timestamp, it count unexpected event into result.
In this case, user 123 has event 1 and event 2 both happened at 100(event_ms), theoretically, it should not count this match because timestamp of event 2 is not greater than event 1.
```sql
SELECT
    uid,
    windowFunnel(50)(toUInt32(event_ms), event_id = 1, event_id = 2) AS level
FROM
    (
    /* emulate test dataset */
    SELECT data.1 event_id, data.2  event_ms, data.3 uid
    FROM (
        SELECT arrayJoin(
        [(1, 100, 123),
        (1, 200, 345),
        (1, 210, 345),
        (2, 100, 123),
        (2, 250, 234),
        (2, 230, 345),
        (2, 240, 345)]) data)
    )
WHERE
    event_ms >= 0
    AND event_ms <= 300
GROUP BY
    uid;

Results:
┌─uid─┬─level─┐
│ 123 │     2 │
│ 234 │     0 │
│ 345 │     2 │
└─────┴───────┘
```

Test 3: running `windowFunnel` on two same event count all events twice:
```sql
SELECT
    uid,
    windowFunnel(50)(toUInt32(event_ms), event_id = 1, event_id = 1) AS level
FROM
    (
    /* emulate test dataset */
    SELECT data.1 event_id, data.2  event_ms, data.3 uid
    FROM (
        SELECT arrayJoin(
        [(1, 100, 123),
        (1, 200, 345)]) data)
    )
WHERE
    event_ms >= 0
    AND event_ms <= 300
GROUP BY
    uid;

Results:
┌─uid─┬─level─┐
│ 123 │     2 │
│ 345 │     2 │
└─────┴───────┘
```

**How to reproduce**: run SQL query pasted above
* Which ClickHouse server version to use: 21.2.3.15
* Which interface to use, if matters
* Non-default settings, if any
* `CREATE TABLE` statements for all tables involved
* Sample data for all these tables, use [clickhouse-obfuscator](https://github.com/ClickHouse/ClickHouse/blob/master/programs/obfuscator/Obfuscator.cpp#L42-L80) if necessary
* Queries to run that lead to unexpected result

**Expected behavior**
When searching for qualified event chain, it should satisfy: 
**timestamp of first event < timestamp of second event <= timestamp of first event + window**

I guess currently it follows:
timestamp of first event <= timestamp of second event <= timestamp of first event + window

With that, Test 2 will generate results:
```
┌─uid─┬─level─┐
│ 123 │     1 │
│ 234 │     0 │
│ 345 │     2 │
└─────┴───────┘
```

Test 3 will generate results:
```
┌─uid─┬─level─┐
│ 123 │     1 │
│ 345 │     1 │
└─────┴───────┘

```

**Error message and/or stacktrace**

**Additional context**

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Added `strict_once` mode to aggregate function `windowFunnel` to  avoid counting one event several times in case it matches multiple conditions, close https://github.com/ClickHouse/ClickHouse/issues/21835

Close https://github.com/ClickHouse/ClickHouse/pull/69302

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit e9b57820fca87ff99a91c8fb6556d5dd9ca3dca8 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69738/e9b57820fca87ff99a91c8fb6556d5dd9ca3dca8/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/69738/e9b57820fca87ff99a91c8fb6556d5dd9ca3dca8/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/AggregateFunctions/AggregateFunctionWindowFunnel.cpp [hunk 1]
```diff
@@ -272,21 +356,34 @@ class AggregateFunctionWindowFunnel final
...
 
     void merge(AggregateDataPtr __restrict place, ConstAggregateDataPtr rhs, Arena *) const override
     {
         this->data(place).merge(this->data(rhs));
     }
 
-    void serialize(ConstAggregateDataPtr __restrict place, WriteBuffer & buf, std::optional<size_t> /* version */) const override
+    /// Versioning for serialization
+    /// Version 1 supports deduplication of the same event several times
+    static constexpr auto MIN_REVISION_FOR_V1 = 54470;
```
Comment (vdimir) [inline src/AggregateFunctions/AggregateFunctionWindowFunnel.cpp:371]: TODO: bump revision if PR is not merged before 24.9 release

File: src/AggregateFunctions/AggregateFunctionWindowFunnel.cpp [hunk 2]
```diff
@@ -169,55 +210,110 @@ class AggregateFunctionWindowFunnel final
...
                     break;
                 else
                     continue;
             }
             else if (event_idx == 0)
             {
-                events_timestamp[0] = std::make_pair(timestamp, timestamp);
-                first_event = true;
+                if (!strict_once)
+                    /// Do not keep all sequences, when mode is disabled
```
Comment (CurtizJ) [inline src/AggregateFunctions/AggregateFunctionWindowFunnel.cpp:250]: Maybe implement separate method for `strict_once` that uses the sequence? It looks like it still gives overhead for other modes that affects [performance](https://s3.amazonaws.com/clickhouse-test-reports/69738/29ed3f4daf1d6cc0a5a80e6c848a1f1e9fb2697c/performance_comparison__release__[3_4]/report.html).

**Comment by vdimir** [comment]
I tried to create specializations based on templates like https://github.com/ClickHouse/ClickHouse/blob/154046ae49c1e8f3c3b76796997170effc83d546/src/AggregateFunctions/AggregateFunctionWindowFunnel.cpp#L219-L252

The result was better, but there was still some performance loss somewhere.
So, finally I implemented it using two separate aggregate data structures and two separate `getEventLevel*` methods. This approach might actually be easier to read than `if constexpr` and following different struct specializations.

**Review by CurtizJ — APPROVED** [review]

**Pull request merged**